src/main/java/com/varukha/onlinebookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,104 @@
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Override
+    public void createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto createCartItem(CreateCartItemRequestDto requestDto) {
+        CartItem existedCartItem = getCartItemByBookId(requestDto.getBookId());
+        if (existedCartItem == null || !existedCartItem.getBook().getId()
+                .equals(requestDto.getBookId())) {
+            return createNewCartItem(requestDto);
+        } else {
+            return updateCartItemQuantityWhenSave(existedCartItem, requestDto.getQuantity());
+        }
+    }
+
+    @Override
+    public CartItem getCartItemById(Long id) {
+        return cartItemRepository.findById(id).orElse(null);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCartByUserId(Long id) {
+        return shoppingCartRepository.findByUserId(getUserId())
+                .map(shoppingCartMapper::toDto)
+                .orElseThrow();
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        cartItemRepository.deleteById(id);
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto updateCartItemQuantity(Long id,
+                                              UpdateCartItemRequestDto cartItemRequestDto) {
+        CartItem cartItemById = getCartItemById(id);
+        cartItemById.setQuantity(cartItemRequestDto.getQuantity());
+        CartItem savedCartItem = cartItemRepository.save(cartItemById);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private CartItemDto updateCartItemQuantityWhenSave(CartItem existingCartItem, int quantity) {
+        existingCartItem.setQuantity(existingCartItem.getQuantity() + quantity);
+        CartItem savedCartItem = cartItemRepository.save(existingCartItem);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private CartItemDto createNewCartItem(CreateCartItemRequestDto requestDto) {
+        CartItem cartItem = new CartItem();
+        cartItem.setShoppingCart(shoppingCartRepository
+                .findByUserId(getUserId())
+                .orElseThrow());
+        cartItem.setBook(bookRepository
+                .findById(requestDto.getBookId()).orElseThrow());
+        cartItem.setQuantity(requestDto.getQuantity());
+        CartItem savedCartItem = cartItemRepository.save(cartItem);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private Long getUserId() {
```

Comment:
Why is this method in ShoppingCartService? It's better to put it in the UserService or Authentication service at all. And rename it to `getAuthenticatedUserId` or something like that

Reply:
Completely agree with you, but I did it to resolve circular dependence of bins problem.